Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,239 @@
+3. `ToString()` allocates yet another managed array **{4}**
+
+That is *{4}* allocations to get a string out of native code. The best you can do to limit this is to reuse the `StringBuilder`
+in another call but this still only saves *1* allocation. It is much better to use and cache a native buffer- you can then
+get down to just the allocation for the `ToString()` on subsequent calls.
+
+The other issue with `StringBuilder` is that it always copies the return buffer back up to the first null. If the passed back string isn't terminated or is a double-null-terminated string your P/Invoke is incorrect at best.
+
+If you *do* use `StringBuilder` one last gotcha is that the capacity does **not** include a hidden null which is always accounted for in interop. It is pretty common for people to get this wrong as most APIs want the size of the buffer *including* the null. This can result in wasted/unnecessary allocations.
+
+**[USE]** Char arrays from `ArrayPool` or `StringBuffer`.
+
+[Default Marshalling for Strings](https://msdn.microsoft.com/en-us/library/s9ts558h.aspx "MSDN")
+
+> ### Windows Specific
+
+> For `[Out]` strings the CLR will use `CoTaskMemFree` by default to free strings or `SysStringFree` for strings that are marked
+as `UnmanagedType.BSTR`.
+
+> **For most APIs with an output string buffer:**
+
+> The passed in character count must include the null. If the returned value is less than the passed in character count the call has succeeded and the value is the number of characters *without* the trailing null. Otherwise the count is the required size of the buffer *including* the null character.
+
+> - Pass in 5, get 4: The string is 4 characters long with a trailing null.
+> - Pass in 5, get 6: The string is 5 characters long, need a 6 character buffer to hold the null.
+
+> [Windows Data Types for Strings](http://msdn.microsoft.com/en-us/library/dd374131.aspx "MSDN")
+
+Booleans
+--------
+
+Booleans are easy to mess up. The default marshalling for P/Invoke is as the Windows type `BOOL`, where it is a 4 byte value. `BOOLEAN`, however, is a *single* byte. This can lead to hard to track down bugs as half the return value will be discarded, which will only *potentially* change the result. For `BOOLEAN` attributing `bool` with either `[MarshalAs(UnmanagedType.U1)]` or `[MarshalAs(UnmanagedType.I1)]` will work as `TRUE` is defined as `1` and `FALSE` is defined as `0`. `U1` is technically more correct as `BOOLEAN` is defined as an `unsigned char`.
+
+For COM (`VARIANT_BOOL`) the type is `2` bytes where true is `-1` and false is `0`. Marshalling uses this by default for bool in COM calls (`UnmanagedType.VariantBool`).
+
+[Default Marshalling for Boolean Types](https://msdn.microsoft.com/en-us/library/t2t3725f.aspx "MSDN")  
+
+Guids
+-----
+
+Guids are usable directly in signatures. When passed by ref, however, they should *not* be marked as `out` or `ref`. Instead the parameter should get the `[MarshalAs(UnmanagedType.LPStruct)]` attribute.
```

Comment:
I do not understand this. You can fine number of PInvoke signatures that have `ref Guid` in CoreCLR and corefx; and they work just fine.